presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -379,14 +385,72 @@ private PrestoSparkTaskSourceRdd createTaskSourcesRdd(
+    @VisibleForTesting
+    public static SetMultimap<Integer, ScheduledSplit> assignSourceDistributionSplits(Session session, List<ScheduledSplit> splits)
     {
-        int taskCount = getSparkInitialPartitionCount(session);
-        checkArgument(taskCount > 0, "taskCount must be greater then zero: %s", taskCount);
         ImmutableSetMultimap.Builder<Integer, ScheduledSplit> result = ImmutableSetMultimap.builder();
-        for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
-            result.put(splitIndex % taskCount, splits.get(splitIndex));
+
+        long maxSplitsSizeInBytesPerPartition = getMaxSplitsDataSizePerSparkPartition(session).toBytes();
+        checkArgument(maxSplitsSizeInBytesPerPartition > 0,
+                "maxSplitsSizeInBytesPerPartition must be greater than zero: %s", maxSplitsSizeInBytesPerPartition);
+
+        boolean splitsDataSizeAvailable = splits.stream()
+                .allMatch(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().isPresent());
+        boolean autoTuneInitialPartitionCount = isSparkInitialPartitionCountAutoTuneEnabled(session) && splitsDataSizeAvailable;
+
+        int initialPartitionCount;
+        if (!autoTuneInitialPartitionCount) {
+            initialPartitionCount = getSparkInitialPartitionCount(session);
+            checkArgument(initialPartitionCount > 0, "initialPartitionCount must be greater then zero: %s", initialPartitionCount);
+        }
+        else {
+            long totalSizeInBytes = splits.stream()
+                    .mapToLong(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong())
+                    .sum();
+            initialPartitionCount = max(1, toIntExact((totalSizeInBytes + maxSplitsSizeInBytesPerPartition - 1) / maxSplitsSizeInBytesPerPartition));
+        }
+
+        if (!splitsDataSizeAvailable) {
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                result.put(splitIndex % initialPartitionCount, splits.get(splitIndex));
+            }
+        }
+        else {
+            splits.sort((ScheduledSplit o1, ScheduledSplit o2) -> {
+                long size1 = o1.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong();
+                long size2 = o2.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong();
+                return size1 == size2 ? 0 : size1 > size2 ? -1 : 1;
+            });
+
+            PriorityQueue<SparkPartition> queue = new PriorityQueue();
+            int partitionCount = 0;
+
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                int partitionId;
+                long splitSizeInBytes = splits.get(splitIndex).getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong();
+
+                if (partitionCount >= initialPartitionCount &&
```

Comment:
So if we look at this if-statement, it's like this:
```
    if (queue is not empty and some other cnoditions) {
        Do something
        Add the partition into the queue
    }
    else {
         // queue is empty
         Do something, but doesn't append anything into the queue
    }
```

As the queue is empty at the beginning, it looks like nothing will be added into the queue. Is there anything I am missing in the code flow? ~ -- or maybe just debug through it to see if the "else" branch will ever get triggered?

Reply:
Yes, I forgot to add it back to the queue, as a result every split is assigned to a new partition.  Unit test is updated with explicit examples to cover this case.